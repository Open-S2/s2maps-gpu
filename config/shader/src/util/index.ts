export * from './bundle';
export * from './tree';
