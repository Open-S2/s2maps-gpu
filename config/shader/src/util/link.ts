import { Tree } from '@lezer/common';
import { ShaderModule, ParsedBundle, ParsedModule, ParsedModuleCache, ShaderDefine, ImportRef, RefFlags as RF } from '../types';
import { VIRTUAL_BINDINGS } from '../constants';

import { bindBundle, bindModule } from './bind';
import { toBundle, getBundleKey } from './bundle';
import { resolveShakeOps } from './shake';
import mapValues from 'lodash/mapValues';

export type Linker = (
  source: ParsedBundle,
  libraries?: Record<string, ShaderModule>,
) => string;

export type LoadModuleWithCache = (
  code: string,
  name: string,
  entry?: string,
  cache?: ParsedModuleCache | null,
) => ParsedModule;

export type GetPreambles = () => string[];

export type GetRenames = (
  defines?: Record<string, ShaderDefine> | null,
) => Map<string, string>;

export type DefineConstants = (
  defines: Record<string, ShaderDefine>
) => string;

export type DefineEnables = (
  enabled: string[],
) => string;

export type RewriteUsingAST = (
  code: string,
  tree: Tree,
  rename: Map<string, string>,
  shake?: number[] | null,
  optionals?: Set<string> | null,
) => string;

const NO_LIBS: Record<string, ShaderModule> = {};

// Link a source module with static modules and dynamic links.
export const makeLinkCode = (
  linker: Linker,
  loadModuleWithCache: LoadModuleWithCache,
  defaultCache: ParsedModuleCache,
) => (
  code: string,
  libraries: Record<string, string> = {},
  links?: Record<string, string | null> | null,
  defines?: Record<string, ShaderDefine> | null,
  cache: ParsedModuleCache | null = defaultCache,
) => {
  const main = loadModuleWithCache(code, 'main', undefined, cache);

  const parsedLibraries = mapValues(libraries, (code: string, name: string) => loadModuleWithCache(code, name, undefined, cache));
  const parsedLinks = mapValues(links, (code: string, name: string) =>
    (
      code != null
      ? loadModuleWithCache(code, name.split(':')[0], undefined, cache)
      : null
    ) as ShaderModule | null
  ) as any;

  const bundle = bindModule(main, parsedLinks, defines);
  return linker(bundle, parsedLibraries);
};

// Link a bundle of parsed module + libs, dynamic links
export const makeLinkBundle = (
  linker: Linker,
) => (
  source: ShaderModule,
  links?: Record<string, ShaderModule | null>,
  defines?: Record<string, ShaderDefine> | null,
) => {
  let bundle = toBundle(source);
  if (links || defines) bundle = bindBundle(bundle, links, defines);

  return linker(bundle);
};

// Link a bundle of parsed module + libs, dynamic links
export const makeLinkModule = (
  linker: Linker,
) => (
  source: ParsedModule,
  libraries: Record<string, ShaderModule> = NO_LIBS,
  links?: Record<string, ShaderModule | null>,
  defines?: Record<string, ShaderDefine> | null,
) => {
  let bundle = toBundle(source);
  if (links || defines) bundle = bindBundle(bundle, links, defines);

  return linker(bundle, libraries);
};

// Make a shader linker with injectable language rules
export const makeLinker = (
  getPreambles: GetPreambles,
  getRenames: GetRenames,
  defineConstants: DefineConstants,
  defineEnables: DefineEnables,
  rewriteUsingAST: RewriteUsingAST,
) => (
  source: ShaderModule,
  libraries: Record<string, ShaderModule> = NO_LIBS,
) => {
  const bundle = toBundle(source);
  const main = getBundleKey(source);

  const {bundles, exported, imported, aliased} = loadBundlesInOrder(bundle, libraries);
  const program = getPreambles();

  // Gather defines/enables
  const defs: Record<string, ShaderDefine> = {};
  const enabled: string[] = [];
  for (const {defines, module: {table: {enables}}} of bundles) {
    if (enables) for (let k of enables) enabled.push(k);
    if (defines) for (let k in defines) defs[k] = defines[k];
  }

  const en = defineEnables(enabled)
  if (en.length) program.push(en);

  const staticRename = getRenames(defs);
  const def = defineConstants(defs);
  if (def.length) program.push(def);

  // Namespace by module key
  const namespaces = new Map<number, string>();

  // Track symbols in global namespace 
  const exists = new Set<string>();
  const visible = new Set<string>();
  const fixed = new Map<string, string>();

  // Track link signatures
  const signatures = new Map<string, any>();
  const infers = new Map<string, string>();

  let hasBoundVirtuals = false;

  for (const bundle of bundles) {
    const {module, defines} = bundle;
    const {name, code, tree, table, shake, virtual} = module;
    const {globals, symbols, visibles, externals, modules, exports: exp} = table;

    const key = getBundleKey(bundle);
    const importMap = imported.get(key);
    const aliasMap = aliased.get(key);

    let optionals: Set<string> | null = null;

    // Namespace all non-global symbols outside main module
    let scope = '';
    const rename = new Map<string, string>();
    if (key !== main) {
      const namespace = virtual?.namespace;
      const ns = reserveNamespace(key, namespaces, namespace);
      scope = ns;

      if (symbols) for (const name of symbols) rename.set(name, ns + name);
      if (globals) for (const name of globals) {
        rename.set(name, name);
        fixed.set(ns + name, name);
      }
      if (visibles) for (const name of visibles) visible.add(rename.get(name)!);
      for (const name of rename.values()) exists.add(name);

      // Gather all exported signatures for type inference
      if (exp) for (const {flags, func} of exp) {
        if (func && (flags & RF.Exported)) {
          const {name} = func;
          signatures.set(ns + name, func);
        }
      }
    }

    // Replace imported symbol names with target
    if (modules) for (const {name: module, imports} of modules) {
      const key = importMap!.get(module)!;
      const ns = namespaces.get(key);

      for (const {name, imported} of imports) {
        let imp = ns + imported;
        if (fixed.has(imp)) imp = fixed.get(imp)!;
        if (!exists.has(imp)) {
          console.warn(`Import ${name} from '${module}' does not exist`);
          debugger;
        }
        else if (!visible.has(imp)) console.warn(`Import ${name} from '${module}' is private`);
        rename.set(name, imp);
        infers.set(scope + name, imp);
      }
    }

    // Replace imported function prototype names with target
    if (externals) for (const {flags, func, variable, struct} of externals) if (func ?? variable ?? struct) {
      const {name, inferred} = func ?? variable ?? struct;
      const key = importMap?.get(name)!;
      const ns = namespaces.get(key);

      const resolved = aliasMap?.get(name) ?? name;
      if ((ns === undefined) && (flags & RF.Optional)) {
        if (!optionals) optionals = new Set();
        optionals!.add(name);
        continue;
      }

      let imp = ns + resolved;
      if (fixed.has(imp)) imp = fixed.get(imp)!;
      if (!exists.has(imp)) {
        console.warn(`Link ${name}:${resolved} does not exist`);
        debugger;
      }
      else if (!visible.has(imp)) console.warn(`Link ${name}:${resolved} is private`);
      rename.set(name, imp);

      if (inferred) {
        const sig = signatures.get(imp);
        const {type, parameters} = sig;
        for (const {name, at} of inferred) {
          const resolved = at < 0 ? type : parameters[at];

          let imp = ns + (resolved.type ?? resolved.name ?? resolved);
          let i = imp;
          while (i = infers.get(imp)) { imp = i; }

          rename.set(name, imp);
          infers.set(scope + name, imp);
        }
      }
    }

    // Copy over static renames
    for (let k of staticRename.keys()) rename.set(k, staticRename.get(k)!);

    if (name === VIRTUAL_BINDINGS) hasBoundVirtuals = true;

    if (virtual) {
      const {uniforms, storages, textures} = virtual;
      if ((uniforms || storages || textures) && (!hasBoundVirtuals)) {
        const id = code.replace('@virtual ', '');
        throw new Error(`Virtual module ${id} has unresolved data bindings`);
      }

      // Emit virtual module in target namespace,
      // with dynamically assigned binding slots.
      const ns = namespaces.get(key)!;
      const recode = virtual.render(ns, rename, virtual.bindingBase, virtual.volatileBase);
      program.push(recode);
    }
    else if (tree) {
      // Shake tree ops based on which symbols were exported
      const keep = exported.get(key);
      const ops = shake && keep ? resolveShakeOps(shake, keep, symbols) : null;

      // Rename symbols using AST while tree shaking
      const recode = rewriteUsingAST(code, tree, rename, ops, optionals);
      program.push(recode);
    }
    else {
      // Static include
      program.push(code);
    }
  }

  const code = program.join("\n");
  return code;
};

// Load all references from a tree of bundles
// while gathering info about what's exported (for tree shaking).
export const loadBundlesInOrder = (
  bundle: ParsedBundle,
  libraries: Record<string, ShaderModule> = {},
): {
  bundles: ParsedBundle[]
  exported: Map<number, Set<string>>,
  imported: Map<number, Map<string, number>>,
  aliased: Map<number, Map<string, string>>,
} => {
  const graph = new Map<number, number[]>();
  const seen  = new Set<number>();
  const hoist = new Set<number>();

  const exported = new Map<number, Set<string>>();
  const imported = new Map<number, Map<string, number>>();
  const aliased  = new Map<number, Map<string, string>>();

  const out: ParsedBundle[] = [];

  const {module} = bundle;
  const {name, entry} = module;
  const key = getBundleKey(bundle);
  exported.set(key, new Set([entry ?? 'main']));

  // Traverse graph starting from source
  const queue = [{key, name, chunk: bundle as ShaderModule}];
  seen.add(key);

  const getContext = (m: ParsedModule) => {
    const {name, code} = m;
    if (name.match(/^_[A-Z]{2}_/) && code.match(/^@/)) return code.replace(/^@/, '');
    return name;
  }

  while (queue.length) {
    const next = queue.shift()!;
    const {key, name, chunk} = next;
    if (chunk == null) throw new Error(`Module '${name}' is undefined`);

    const bundle = toBundle(chunk);
    const {module, libs, links: linkDefs} = bundle;
    const {table: {modules, externals}} = module;
    const deps = [] as number[];

    const [links, aliases] = parseLinkAliases(linkDefs);

    // Static renames and imports for this module instance
    let aliasMap: Map<string, string> | null = null;
    let importMap: Map<string, number> | null = null;

    // Recurse into imports
    if (modules) for (const {at, name, imports} of modules) {
      const chunk = libs?.[name] ?? libraries[name];
      if (!chunk) throw new Error(`Module '${name}' in ${getContext(module)} is unknown`);

      const key = getBundleKey(chunk);
      if (!seen.has(key)) queue.push({key, name, chunk});
      seen.add(key);
      deps.push(key);

      if (!importMap) importMap = new Map();
      importMap!.set(name, key);

      if (at < 0) hoist.add(key);

      let list = exported.get(key);
      if (!list) exported.set(key, list = new Set());
      imports.forEach((i: ImportRef) => list!.add(i.imported));
    }

    // Recurse into links
    if (externals) for (const {flags, func, variable, struct} of externals) if (func ?? variable ?? struct) {
      const {name} = func ?? variable ?? struct;
      const chunk = links[name];
      if (!chunk) {
        if (flags & RF.Optional) {
          continue;
        }
        throw new Error(`Link '${name}' in ${getContext(module)} is not linked`);
      }

      const key = getBundleKey(chunk);
      if (!seen.has(key)) queue.push({key, name, chunk});
      seen.add(key);
      deps.push(key);

      if (!importMap) importMap = new Map();
      importMap!.set(name, key);

      const entry = chunk.entry ?? (chunk as any).module?.entry;
      const symbol = entry ?? aliases?.get(name) ?? name;

      if (symbol !== name) {
        if (!aliasMap) aliasMap = new Map();
        aliasMap!.set(name, symbol);
      }

      let list = exported.get(key);
      if (!list) exported.set(key, list = new Set());
      list.add(symbol);
    }

    // Build module-to-module dependency graph
    graph.set(key, deps);

    // Store aliases/imports/bundle
    if (aliasMap) aliased.set(key, aliasMap);
    if (importMap) imported.set(key, importMap);

    out.push(bundle);
  }

  // Sort by graph depth
  const order = getGraphOrder(graph, key);
  for (let [k, v] of order.entries()) if (hoist.has(k)) order.set(k, v + 1e5);
  out.sort((a, b) => order.get(getBundleKey(b))! - order.get(getBundleKey(a))! || a.module.name.localeCompare(b.module.name));

  return {
    bundles: out,
    exported,
    imported,
    aliased,
  };
};


// Generate a new namespace
export const reserveNamespace = (
  key: string | number,
  namespaces: Map<any, string>,
  force?: string,
): string => {
  let namespace = force ?? '_' + ('00' + (namespaces.size + 1).toString(36)).slice(-2) + '_';
  namespaces.set(key, namespace);
  return namespace;
}

// Get depth for each item in a graph, so its dependencies resolve correctly
export const getGraphOrder = (
  graph: Map<number, number[]>,
  name: number,
  depth: number = 0,
) => {
  const queue = [{name, depth: 0, path: [name]}];
  const depths = new Map<number, number>();

  while (queue.length) {
    const {name, depth, path} = queue.shift()!;
    depths.set(name, depth);

    const module = graph.get(name);
    if (!module) continue;

    const deps = module.map(name => {
      const i = path.indexOf(name);
      if (i >= 0) throw new Error("Cycle detected in module dependency graph: " + path.slice(i));
      return {name, depth: depth + 1, path: [...path, name]};
    });

    queue.push(...deps);
  }
  return depths;
}

// Parse run-time specified keys `from:to` into a map of aliases
export const parseLinkAliases = <T>(
  links?: Record<string, T>,
): [
  Record<string, T>,
  Map<string, string> | null,
] => {
  const out = {} as Record<string, T>;
  let aliases = null as Map<string, string> | null;

  for (let k in links) {
    const link = links[k] as any;
    if (!link) continue;

    let [name, imported] = k.split(':');
    if (!imported && link.entry != null) imported = link.entry;
    if (!imported && link.module?.entry != null) imported = link.module.entry;

    out[name] = link;
    if (imported) {
      if (!aliases) aliases = new Map<string, string>();
      aliases.set(name, imported);
    }
  }

  return [out, aliases];
}
