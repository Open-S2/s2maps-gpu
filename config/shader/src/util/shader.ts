import { Tree } from '@lezer/common';
import { ASTParser, VirtualTable, SymbolTableT, ParsedBundle, ParsedModule, ParsedModuleCache, CompressedNode } from '../types';
import { formatMurmur53, toMurmur53 } from './hash';
import { PREFIX_VIRTUAL } from '../constants';

const EMPTY_LIST = [] as any[];
const EMPTY_TABLE = {} as any;

// Parse a code module into its in-memory representation
// (AST + symbol table)
export const makeLoadModule = <T extends SymbolTableT = any>(
  parseShader: (code: string) => Tree,
  makeASTParser: (code: string, tree: Tree, name?: string) => ASTParser<T>,
  compressAST: (code: string, tree: Tree, symbols?: string[]) => CompressedNode[],
  decompressAST: (nodes: CompressedNode[], symbols?: string[]) => Tree,
) => (
  code: string,
  name: string = 'main',
  entry?: string,
  compressed: boolean = false,
): ParsedModule => {
  if (code == null) throw new Error(`Shader code ${name} undefined`);
  if (typeof code !== 'string') throw new Error(`Shader code ${name} is not a string`);
  let tree = parseShader(code);

  const astParser = makeASTParser(code, tree, name);
  const table = astParser.getSymbolTable();
  const shake = astParser.getShakeTable(table);

  if (compressed) {
    const {symbols} = table;
    tree = decompressAST(compressAST(code, tree, symbols), symbols);
  }
  const hash = toMurmur53(code);

  return bindEntryPoint({name, code, hash, table, shake, tree}, entry);
}

// Use cache to load modules
export const makeLoadModuleWithCache = (
  loadModule: (code: string, name?: string, entry?: string, compressed?: boolean) => ParsedModule,
  defaultCache: ParsedModuleCache,
) => (
  code: string,
  name?: string,
  entry?: string,
  cache: ParsedModuleCache | null = defaultCache,
): ParsedModule => {
  if (!cache) return loadModule(code, name, entry, true);

  const hash = toMurmur53(code);
  const cached = cache.get(hash);
  if (cached) {
    return bindEntryPoint(cached, entry);
  }
  
  const module = loadModule(code, name, undefined, true);
  cache.set(hash, module);
  return bindEntryPoint(module, entry);
}

// Load a static (inert) module
export const loadStaticModule = (code: string, name: string, entry?: string) => {
  const hash = toMurmur53([code, entry]);
  return ({ name, code, hash, table: EMPTY_TABLE });
}

// Load a virtual (generated) module
export const loadVirtualModule = <T extends SymbolTableT = any>(
  virtual: VirtualTable,
  initTable: Partial<T> = EMPTY_TABLE,
  entry?: string,
  hash?: number,
  code?: string,
  key?: number,
) => {
  let symbols = initTable.symbols ?? EMPTY_LIST;

  code = code ?? `@virtual [${symbols.join(' ')}]`;
  hash = hash ?? toMurmur53(code);
  key  = key  ?? hash;

  const name = `${PREFIX_VIRTUAL}${formatMurmur53(key).slice(0, 6)}`;

  const table = {
    symbols,
    visibles: symbols,
    ...initTable,
  };
  return { name, code, hash, table, entry, virtual, key };
}

// Set entry point of a module, returns new bundle/module.
// Is the same instance as the original (key = old key/hash), so it merges with copies of itself.
// But is structurally different (hash = new key), so differences in links are reflected in the shader hash.
export const bindEntryPoint = <T extends ParsedBundle | ParsedModule>(bundle: T, entry?: string): T => {
  let {key, hash, module, table} = bundle as any;

  table = table ?? module?.table;
  hash = hash ?? module?.hash;

  if (entry == null && table.symbols?.includes('main')) entry = 'main';
  if (entry == null) return bundle;

  const structural = toMurmur53([hash, entry]);
  return {...bundle, entry, hash: structural, key: key ?? hash};
};
