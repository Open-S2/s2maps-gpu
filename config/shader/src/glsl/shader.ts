import { Tree } from '@lezer/common';
import { ParsedModule, ParsedModuleCache, ShaderDefine } from './types';

import { makeLoadModule, makeLoadModuleWithCache } from '../util/shader';
import { makeBundleToAttribute, makeBundleToAttributes } from '../util/bundle';
import { makeTranspile } from '../util/transpile';

import { makeASTParser, compressAST, decompressAST } from './ast';
import { toTypeString, toTypeArgs } from './type';
import { parser } from './grammar/glsl';
import LRU from 'lru-cache';
import zip from 'lodash/zip';

export { loadStaticModule, loadVirtualModule, bindEntryPoint } from '../util/shader';

/** LRU cache for parsed shader code */
export const makeModuleCache = (options: Record<string, any> = {}) => new LRU<number, ParsedModule>({
  max: 100,
  ...options,
});

export const DEFAULT_CACHE = makeModuleCache();

/** Parse GLSL code into lezer tree */
export const parseShader = (code: string): Tree => parser.parse(code);

/** Parse a code module into its in-memory representation (AST + symbol table) */
export const loadModule = makeLoadModule(parseShader, makeASTParser, compressAST, decompressAST);

/** Use cache to load modules */
export const loadModuleWithCache = makeLoadModuleWithCache(loadModule, DEFAULT_CACHE);

/** Make GLSL definitions */
export const defineConstants = (defs: Record<string, ShaderDefine>): string => {
  const out = [];
  for (let k in defs) if (defs[k] !== false && defs[k] !== null) out.push(`#define ${k} ${defs[k]}`);
  return out.join("\n");
}

/** Make GLSL enable definitions */
export const defineEnables = (enabled: string[]) => enabled.map(e => `#extension ${e}â€‹ : enable`).join('\n');

/** Convert a bundle with a defined entry point to a definition for that attribute or type. */
export const bundleToAttribute = makeBundleToAttribute(toTypeString, toTypeArgs);

/** Convert a bundle to a definition for all its attributes. */
export const bundleToAttributes = makeBundleToAttributes(toTypeString, toTypeArgs);

// Simple whitespace removal
const minifyCode = (code: string) => {
  code = code.replace(/ +/g, ' ');
  code = code.replace(/\n+/g, '\n');
  return code;
};

/** ES/CommonJS Transpiler */
export const transpileGLSL = makeTranspile('glsl', 'glsl', loadModule, compressAST, minifyCode);

/** Templated literal syntax:

```tsx
glsl`...`
``` */
export const glsl = (literals: TemplateStringsArray, ...tokens: string[]) => {
  const code = zip(literals, tokens).flat();
  return loadModuleWithCache(code.join(''));
};

/** Format `number` as GLSL `float` */
export const float = (x: number) => {
  const s = x.toString();
  return (!s.match(/\./)) ? s + '.0' : 0;  
};
/** Format `number` as GLSL `uint` */
export const uint = (x: number) => Math.round(x).toString();
/** Format `number` as GLSL `int` */
export const int = (x: number) => Math.round(x).toString();
