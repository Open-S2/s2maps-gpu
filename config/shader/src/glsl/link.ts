import { Tree } from '@lezer/common';

import { defineConstants, defineEnables, loadModule, loadModuleWithCache, DEFAULT_CACHE } from './shader';
import { rewriteUsingAST } from './ast';
import { makeLinker, makeLinkCode, makeLinkBundle, makeLinkModule } from '../util/link';

import { GLSL_VERSION } from './constants';

// Override GLSL version/prefix
let PREAMBLE = `#version ${GLSL_VERSION}`;
export const setPreamble = (s: string): string => PREAMBLE = s;
export const getPreamble = (): string => PREAMBLE;
const getPreambles = (): string[] => [getPreamble()];

// No preprocessor renames
const getRenames = () => new Map<string, string>();

export const linker     = makeLinker(getPreambles, getRenames, defineConstants, defineEnables, rewriteUsingAST);
export const linkBundle = makeLinkBundle(linker);
export const linkModule = makeLinkModule(linker);
export const linkCode   = makeLinkCode(linker, loadModuleWithCache, DEFAULT_CACHE);
