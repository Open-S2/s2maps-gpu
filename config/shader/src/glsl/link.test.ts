import { GLSLModules } from './glsl.test.data';
import { linkCode, linkModule } from './link';
import { loadModule, glsl } from './shader';
import { formatAST } from '../util/tree'; 
import { addASTSerializer } from '../test/snapshot';
import mapValues from 'lodash/mapValues';

const loadedModules = mapValues(GLSLModules, (v, k) => loadModule(v, k, k));

addASTSerializer(expect);

describe("link", () => {
  
  it("links an external", () => {
    
    const code = `
    vec4 getColor();
    void main() {
      vec4 v;
      v.xyz = vec3(1.0, 0.0, 0.0);
      gl_FragColor = getColor();
    }
    `
    
    const getColor = `
    #pragma export
    vec4 getColor() { return vec4(1.0, 0.0, 1.0, 1.0); }
    `
    
    const linked = linkCode(code, {}, {getColor});
    expect(linked).toMatchSnapshot();

  });

  it("links quad vertex", () => {

    const getPosition = glsl`
    #pragma export
    vec4 getPosition(int index) { return vec4(1.0, 0.0, 1.0, 1.0); }
    `

    const getPerspective = glsl`
    #pragma export
    vec4 getPerspective(int index) { return 1.0; }
    `

    const getColor = glsl`
    #pragma export
    vec4 getColor(int index) { return vec4(1.0, 0.0, 1.0, 1.0); }
    `

    const getSize = glsl`
    #pragma export
    float getSize(int index) { return 1.0; }
    `

    const getDepth = glsl`
    #pragma export
    float getDepth(int index) { return 0.5; }
    `

    const module = loadedModules['getQuadVertex'];
    const linked = linkModule(module, loadedModules, {getPosition, getPerspective, getColor, getSize, getDepth});
    expect(linked).toMatchSnapshot();

  });
  
  it("lifts recursive dependency", () => {
    
    const code = `
    #pragma import {getLifted} from 'getLifted'
    #pragma import {getColor1} from 'getColor1'
    void main() {
      gl_FragColor = getColor1();
    }
    `

    const getLifted = `
    // Lifted Code
    #pragma export
    void getLifted() {}
    `

    const getColor1 = `
    #pragma import {getColor2} from 'getColor2'
    #pragma export
    vec4 getColor1() { return getColor2(); }
    `
    
    const getColor2 = `
    #pragma import {getLifted} from 'getLifted'
    #pragma export
    vec4 getColor2() { return vec4(1.0, 0.0, 1.0, 1.0); }
    `
    
    const linked = linkCode(code, {getColor1, getColor2, getLifted});
    expect(linked.indexOf('// Lifted Code')).toBeLessThan(linked.indexOf('getColor2'));
    expect(linked).toMatchSnapshot();

  });
  
  it("tree shakes constants", () => {
    const sub = `
    const vec4 colorUsed = vec4(0.0, 0.1, 0.2, 0.0);
    const vec4 colorNotUsed = vec4(0.0, 0.1, 0.2, 1.0);

    #pragma export
    vec4 getColor() {
      return colorUsed;
    }
    `;

    const main = `
    vec4 getColor();
    void main() {
      vec4 a = getColor();
    }
    `;

    for (let compressed of [false, true]) {
      const modMain = loadModule(main, 'main', undefined, compressed);
      const modSub = loadModule(sub, 'sub', undefined, compressed);

      const getPosition = {...modSub, entry: 'getPosition'};
      const getColor = {...modSub, entry: 'getColor'};

      const linked = linkModule(modMain, {}, {getPosition, getColor});
      expect(linked).toMatchSnapshot();
    }
  })
  
  it("tree shakes around identifiers", () => {

    const sub = `
    float used() { return 1.0; }

    float unused() { return 1.0; }
    
    #pragma export
    vec4 getPosition(int index) { return vec4(used(), 0.0, 1.0, 1.0); }

    #pragma export
    vec4 getColor(int index) { return vec4(used(), 0.0, 1.0, 1.0); }
    `

    const main = `
    vec4 getPosition(int index);
    void main() {
      vec4 a = getPosition(0);
    }
    `

    for (let compressed of [false, true]) {
      const modMain = loadModule(main, 'main', undefined, compressed);
      const modSub = loadModule(sub, 'sub', undefined, compressed);

      const getPosition = {...modSub, entry: 'getPosition'};
      const getColor = {...modSub, entry: 'getColor'};

      const linked = linkModule(modMain, {}, {getPosition, getColor});
      expect(linked).toMatchSnapshot();
    }

  });
  
  it("links same module twice with different entry point", () => {

    const sub = `
    float used() { return 1.0; }
    
    #pragma export
    vec4 getPosition(int index) { return vec4(used(), 0.0, 1.0, 1.0); }

    #pragma export
    vec4 getColor(int index) { return vec4(used(), 0.0, 1.0, 1.0); }
    `

    const main = `
    vec4 getPosition(int index);
    vec4 getColor(int index);
    void main() {
      vec4 a = getPosition(0);
      vec4 b = getColor(0);
    }
    `

    const modMain = loadModule(main, 'main', undefined, false);
    const modSub = loadModule(sub, 'sub', undefined, false);

    const getPosition = {...modSub, entry: 'getPosition'};
    const getColor = {...modSub, entry: 'getColor'};

    const linked = linkModule(modMain, {}, {getPosition, getColor});
    expect(linked).toMatchSnapshot();

  });

  it("links a global across a module", () => {

    const sub1 = `
    #pragma global
    #pragma export
    vec4 getPosition(int index) { return vec4(1.0, 0.0, 1.0, 1.0); }
    `

    const sub2 = `
    void getPosition() {};
    
    #pragma export
    vec4 getColor(int index) {
      getPosition();
      return vec4(1.0, 0.0, 1.0, 1.0);
    }
    `

    const main = `
    vec4 getPosition(int index);
    vec4 getColor(int index);
    void main() {
      vec4 a = getPosition(0);
    }
    `

    const modMain = loadModule(main, 'main');
    const modSub1 = loadModule(sub1, 'sub1');
    const modSub2 = loadModule(sub2, 'sub2');

    const getPosition = {...modSub1, entry: 'getPosition'};
    const getColor = {...modSub2, entry: 'getColor'};

    const linked = linkModule(modMain, {}, {getPosition, getColor});
    expect(linked).toMatchSnapshot();

    expect(linked).toMatch(/vec4 getPosition/);
    expect(linked).not.toMatch(/vec4 _[A-Za-z0-9]{2,}_getPosition/);

    expect(linked).not.toMatch(/void getPosition/);
    expect(linked).toMatch(/void _[A-Za-z0-9]{2,}_getPosition/);

  });
});
