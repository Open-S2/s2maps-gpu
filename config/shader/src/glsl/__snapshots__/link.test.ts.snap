// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`link lifts recursive dependency 1`] = `
"#version 450

    // Lifted Code
    void _01_getLifted() {}
    

    vec4 _02_getColor2() { return vec4(1.0, 0.0, 1.0, 1.0); }
    

    vec4 _03_getColor1() { return _02_getColor2(); }
    

    void main() {
      gl_FragColor = _03_getColor1();
    }
    "
`;

exports[`link links a global across a module 1`] = `
"#version 450

    vec4 getPosition(int index) { return vec4(1.0, 0.0, 1.0, 1.0); }
    

    void _02_getPosition() {};
    
    vec4 _02_getColor(int index) {
      _02_getPosition();
      return vec4(1.0, 0.0, 1.0, 1.0);
    }
    

    vec4 getPosition(int index);
    void main() {
      vec4 a = getPosition(0);
    }
    "
`;

exports[`link links an external 1`] = `
"#version 450

    vec4 _01_getColor() { return vec4(1.0, 0.0, 1.0, 1.0); }
    

    vec4 _01_getColor();
    void main() {
      vec4 v;
      v.xyz = vec3(1.0, 0.0, 0.0);
      gl_FragColor = _01_getColor();
    }
    "
`;

exports[`link links quad vertex 1`] = `
"#version 450
const ivec2 _01_QUAD[] = {
  ivec2(0, 0),
  ivec2(1, 0),
  ivec2(0, 1),
  ivec2(1, 1),
};

ivec2 _01_getQuadIndex(int vertex) {
  return _01_QUAD[vertex];
}

vec2 _01_getQuadUV(int vertex) {
  return vec2(_01_getQuadIndex(vertex));
}


    vec4 _02_getPosition(int index) { return vec4(1.0, 0.0, 1.0, 1.0); }
    

    vec4 _03_getColor(int index) { return vec4(1.0, 0.0, 1.0, 1.0); }
    

    float _04_getSize(int index) { return 1.0; }
    

    float _05_getDepth(int index) { return 0.5; }
    
struct _06_SolidVertex {
  vec4 position;
  vec4 color;
  vec2 uv;
};


layout(set = VIEW_BINDGROUP, binding = VIEW_BINDING) uniform _07_ViewUniforms {
  mat4 projectionMatrix;
  mat4 viewMatrix;
  vec4 viewPosition;
  vec2 viewResolution;
  vec2 viewSize;
  float viewWorldDepth;
  float viewPixelRatio;
} _07_viewUniforms;

vec4 _07_worldToView(vec4 position) {
  return _07_viewUniforms.viewMatrix * position;
}

vec4 _07_viewToClip(vec4 position) {
  return _07_viewUniforms.projectionMatrix * position;
}

vec4 _07_worldToClip(vec4 position) {
  return _07_viewToClip(_07_worldToView(position));
}

float _07_getPerspectiveScale(float w, float f) {
  mat4 m = _07_viewUniforms.projectionMatrix;
  float worldScale = m[1][1] * _07_viewUniforms.viewWorldDepth;
  float clipScale = mix(1.0, worldScale / w, f);
  float pixelScale = clipScale * _07_viewUniforms.viewPixelRatio;
  return pixelScale;
}

vec4 _02_getPosition(int);
vec4 _03_getColor(int);
vec2 _04_getSize(int);
float _05_getDepth(int);

_06_SolidVertex getQuadVertex(int vertexIndex, int instanceIndex) {
  vec4 position = _02_getPosition(instanceIndex);
  vec4 color = _03_getColor(instanceIndex);
  vec2 size = _04_getSize(instanceIndex);
  float depth = _05_getDepth(instanceIndex);

  vec4 center = _07_worldToClip(position);

  vec2 uv = _01_getQuadUV(vertexIndex);
  vec2 xy = uv * 2.0 - 1.0;
  
  // Lerp between fixed size and full perspective.
  float pixelScale = _07_getPerspectiveScale(center.w, depth);
  size *= pixelScale;

  #ifdef HAS_EDGE_BLEED
  xy = xy * (size + 0.5) / size;
  uv = xy * .5 + .5;
  #endif

  center.xy += xy * size * _07_viewUniforms.viewResolution * center.w;

  return _06_SolidVertex(
    center,
    color,
    uv
  );
}"
`;

exports[`link links same module twice with different entry point 1`] = `
"#version 450

    float _01_used() { return 1.0; }
    
    vec4 _01_getPosition(int index) { return vec4(_01_used(), 0.0, 1.0, 1.0); }

    vec4 _01_getColor(int index) { return vec4(_01_used(), 0.0, 1.0, 1.0); }
    

    vec4 _01_getPosition(int index);
    vec4 _01_getColor(int index);
    void main() {
      vec4 a = _01_getPosition(0);
      vec4 b = _01_getColor(0);
    }
    "
`;

exports[`link tree shakes around identifiers 1`] = `
"#version 450

    float _01_used() { return 1.0; }

    vec4 _01_getPosition(int index) { return vec4(_01_used(), 0.0, 1.0, 1.0); }

    

    vec4 _01_getPosition(int index);
    void main() {
      vec4 a = _01_getPosition(0);
    }
    "
`;

exports[`link tree shakes around identifiers 2`] = `
"#version 450

    float _01_used() { return 1.0; }

    vec4 _01_getPosition(int index) { return vec4(_01_used(), 0.0, 1.0, 1.0); }

    

    vec4 _01_getPosition(int index);
    void main() {
      vec4 a = _01_getPosition(0);
    }
    "
`;

exports[`link tree shakes constants 1`] = `
"#version 450

    const vec4 _01_colorUsed = vec4(0.0, 0.1, 0.2, 0.0);
    vec4 _01_getColor() {
      return _01_colorUsed;
    }
    

    vec4 _01_getColor();
    void main() {
      vec4 a = _01_getColor();
    }
    "
`;

exports[`link tree shakes constants 2`] = `
"#version 450

    const vec4 _01_colorUsed = vec4(0.0, 0.1, 0.2, 0.0);
    vec4 _01_getColor() {
      return _01_colorUsed;
    }
    

    vec4 _01_getColor();
    void main() {
      vec4 a = _01_getColor();
    }
    "
`;
